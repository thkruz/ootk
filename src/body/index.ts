export { Celestial } from './Celestial';
export { Earth } from './Earth';
export { Moon } from './Moon';
export { Sun } from './Sun';
