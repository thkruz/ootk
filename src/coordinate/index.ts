export * from './Hill.js';
export * from './ClassicalElements.js';
export * from './EquinoctialElements.js';
export * from './FormatTle.js';
export * from './Geodetic.js';
export * from './ITRF.js';
export * from './J2000.js';
export * from './RelativeState.js';
export * from './RIC.js';
export * from './StateVector.js';
export * from './TEME.js';
export * from './Tle.js';
