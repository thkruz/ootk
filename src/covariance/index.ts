export * from './CovarianceSample.js';
export * from './StateCovariance.js';
