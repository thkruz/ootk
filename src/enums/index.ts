export { CatalogSource } from './CatalogSource.js';
export { CommLink } from './CommLink.js';
export { OrbitRegime } from './OrbitRegime.js';
export { Sgp4OpsMode } from './Sgp4OpsMode.js';
export { AngularDiameterMethod } from './AngularDiameterMethod.js';
export { AngularDistanceMethod } from './AngularDistanceMethod.js';
export { PassType } from './PassType.js';
