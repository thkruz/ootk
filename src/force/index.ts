export { EarthGravity } from './EarthGravity.js';
export { Force } from './Force.js';
export { ForceModel } from './ForceModel.js';
export { Gravity } from './Gravity.js';
export { SolarRadiationPressure } from './SolarRadiationPressure.js';
export { ThirdBodyGravity } from './ThirdBodyGravity.js';
export { Thrust } from './Thrust.js';

