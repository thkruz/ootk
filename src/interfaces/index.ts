export type { GroundPositionParams } from './GroundPositionParams.js';
export type { OptionsParams } from './OptionsParams.js';
export type { TimeVariables } from './TimeVariables.js';
export type { StarObjectParams } from './StarObjectParams.js';
export type { ClassicalElementsParams } from './ClassicalElementsParams.js';
export type { EquinoctialElementsParams } from './EquinoctialElementsParams.js';
export type { ScenarioParams } from '../objects/Scenario.js';
export type { BaseObjectParams } from '../objects/base/BaseObject.js';
export type { GroundObjectParams } from '../objects/base/GroundObject.js';
export type { FacilityParams } from '../objects/base/Facility.js';
export type { SensorParams } from '../objects/attachable/SensorAdv.js';
export type { RectangularSensorParams } from '../objects/attachable/RectangularSensor.js';
