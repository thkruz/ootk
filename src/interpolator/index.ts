export { ChebyshevCoefficients } from './ChebyshevCoefficients.js';
export { ChebyshevInterpolator } from './ChebyshevInterpolator.js';
export { CubicSpline } from './CubicSpline.js';
export { CubicSplineInterpolator } from './CubicSplineInterpolator.js';
export { Interpolator } from './Interpolator.js';
export { LagrangeInterpolator } from './LagrangeInterpolator.js';
export { StateInterpolator } from './StateInterpolator.js';
export { VerletBlendInterpolator } from './VerletBlendInterpolator.js';

