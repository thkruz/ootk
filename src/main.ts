/**
 * @author @thkruz Theodore Kruczek
 * @description Orbital Object ToolKit (ootk) is a collection of tools for working
 * with satellites and other orbital objects.
 * @license AGPL-3.0-or-later
 * @copyright (c) 2020-2024 Theodore Kruczek
 *
 * Many of the classes are based off of the work of @david-rc-dayton and his
 * Pious Squid library (https://github.com/david-rc-dayton/pious_squid).
 *
 * Orbital Object ToolKit is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * Orbital Object ToolKit is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with
 * Orbital Object ToolKit. If not, see <http://www.gnu.org/licenses/>.
 */

export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './time/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
export * from './operations/operations.js';
export { BaseObject, Sensor } from './objects/index.js';
export { Earth } from './body/index.js';
export * from './coordinate/index.js';
export * from './observation/index.js';
export * from './data/DataHandler.js';
export * from './sgp4/index.js';
export * from './objects/index.js';
export * from './interpolator/index.js';
export * from './force/index.js';
export * from './maneuver/index.js';
export * from './optimize/index.js';
export * from './orbit_determination/index.js';
export * from './body/index.js';
export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './coordinate/index.js';
export * from './operations/index.js';
export * from './force/index.js';
export * from './propagator/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
