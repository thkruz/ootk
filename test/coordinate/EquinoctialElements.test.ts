import { EquinoctialElements, EpochUTC, Kilometers, Radians } from '../../src/main';

describe('ClassicalElements', () => {
  const epoch = EpochUTC.fromDateTime(new Date('2024-01-14T14:39:39.914Z'));
  let elements: EquinoctialElements;
  let elements2: EquinoctialElements;

  beforeEach(() => {
    elements = new EquinoctialElements({
      epoch,
      a: 6935.754028093152 as Kilometers,
      h: 0.000647924735354646,
      k: -0.00023984363760690404,
      p: 0.06877966401254976,
      q: 0.38775089544898517,
      lambda: 7.675800884962325 as Radians,
      mu: 398600.4415,
      I: 1,
    });
    elements2 = new EquinoctialElements({
      epoch,
      a: 6937.389795521736 as Kilometers,
      h: 0.0008140324591775426,
      k: 0.00025609012211642304,
      p: 0.09761971480018718,
      q: 0.3198440470005213,
      lambda: 7.224398200203262 as Radians,
      mu: 398600.4415,
      I: -1,
    });
  });

  // toString
  it('should return a string representation of the EquinoctialElements object', () => {
    expect(elements.toString()).toEqual(
      [
        '[EquinoctialElements]',
        `  Epoch: ${epoch}`,
        `  a: ${elements.a} km`,
        `  h: ${elements.h}`,
        `  k: ${elements.k}`,
        `  p: ${elements.p}`,
        `  q: ${elements.q}`,
        `  lambda: ${elements.lambda} rad`,
      ].join('\n'),
    );
  });

  // semiMajorAxis
  it('should return the semi-major axis', () => {
    expect(elements.semimajorAxis).toMatchSnapshot();
    expect(elements2.semimajorAxis).toMatchSnapshot();
  });

  // meanLongitude
  it('should return the mean longitude', () => {
    expect(elements.meanLongitude).toMatchSnapshot();
    expect(elements2.meanLongitude).toMatchSnapshot();
  });

  // meanMotion
  it('should return the mean motion', () => {
    expect(elements.meanMotion).toMatchSnapshot();
    expect(elements2.meanMotion).toMatchSnapshot();
  });

  // retrogradeFactor
  it('should return the retrograde factor', () => {
    expect(elements.retrogradeFactor).toMatchSnapshot();
    expect(elements2.retrogradeFactor).toMatchSnapshot();
  });

  // isPrograde
  it('should return true if the orbit is prograde', () => {
    expect(elements.isPrograde()).toMatchSnapshot();
    expect(elements2.isPrograde()).toMatchSnapshot();
  });

  // isRetrograde
  it('should return true if the orbit is retrograde', () => {
    expect(elements.isRetrograde()).toMatchSnapshot();
    expect(elements2.isRetrograde()).toMatchSnapshot();
  });

  // period
  it('should return the period', () => {
    expect(elements.period).toMatchSnapshot();
    expect(elements2.period).toMatchSnapshot();
  });

  // revsPerDay
  it('should return the revs per day', () => {
    expect(elements.revsPerDay).toMatchSnapshot();
    expect(elements2.revsPerDay).toMatchSnapshot();
  });

  // toClassicalElements
  it('should return the ClassicalElements object', () => {
    expect(elements.toClassicalElements()).toMatchSnapshot();
    expect(elements2.toClassicalElements()).toMatchSnapshot();
  });

  // toPositionVelocity
  it('should return the PositionVelocity object', () => {
    expect(elements.toPositionVelocity()).toMatchSnapshot();
    expect(elements2.toPositionVelocity()).toMatchSnapshot();
  });
});
