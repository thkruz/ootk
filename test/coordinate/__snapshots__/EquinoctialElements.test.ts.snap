// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClassicalElements should return the ClassicalElements object 1`] = `
ClassicalElements {
  "argPerigee": 1.7497725502749217,
  "eccentricity": 0.0006908917666211547,
  "epoch": EpochUTC {
    "posix": 1705243179.914,
  },
  "inclination": 0.7503068041317155,
  "mu": 398600.4415,
  "rightAscension": 0.17555503341584255,
  "semimajorAxis": 6935.754028093152,
  "trueAnomaly": -0.5334142966624242,
}
`;

exports[`ClassicalElements should return the ClassicalElements object 2`] = `
ClassicalElements {
  "argPerigee": 1.5622345094945798,
  "eccentricity": 0.0008533645148705458,
  "epoch": EpochUTC {
    "posix": 1705243179.914,
  },
  "inclination": 2.4961547762752434,
  "mu": 398600.4415,
  "rightAscension": 0.296230054374017,
  "semimajorAxis": 6937.389795521736,
  "trueAnomaly": -0.32533674963545683,
}
`;

exports[`ClassicalElements should return the PositionVelocity object 1`] = `
Object {
  "position": Vector3D {
    "x": 1538.2233358428962,
    "y": 5102.261204021967,
    "z": 4432.634965003576,
  },
  "velocity": Vector3D {
    "x": -7.3415189093793,
    "y": 0.6516718453998642,
    "z": 1.7933882499862026,
  },
}
`;

exports[`ClassicalElements should return the PositionVelocity object 2`] = `
Object {
  "position": Vector3D {
    "x": 3699.9682301219336,
    "y": -4340.514588573352,
    "z": 3939.5098844969084,
  },
  "velocity": Vector3D {
    "x": -6.276352257339577,
    "y": -3.9904909019658032,
    "z": 1.4943961405848338,
  },
}
`;

exports[`ClassicalElements should return the mean longitude 1`] = `7.675800884962325`;

exports[`ClassicalElements should return the mean longitude 2`] = `7.224398200203262`;

exports[`ClassicalElements should return the mean motion 1`] = `0.0010930206413363934`;

exports[`ClassicalElements should return the mean motion 2`] = `0.0010926340790517568`;

exports[`ClassicalElements should return the period 1`] = `95.80766470395568`;

exports[`ClassicalElements should return the period 2`] = `95.8415604339752`;

exports[`ClassicalElements should return the retrograde factor 1`] = `1`;

exports[`ClassicalElements should return the retrograde factor 2`] = `-1`;

exports[`ClassicalElements should return the revs per day 1`] = `15.030112720621878`;

exports[`ClassicalElements should return the revs per day 2`] = `15.02479710763901`;

exports[`ClassicalElements should return the semi-major axis 1`] = `6935.754028093152`;

exports[`ClassicalElements should return the semi-major axis 2`] = `6937.389795521736`;

exports[`ClassicalElements should return true if the orbit is prograde 1`] = `true`;

exports[`ClassicalElements should return true if the orbit is prograde 2`] = `false`;

exports[`ClassicalElements should return true if the orbit is retrograde 1`] = `false`;

exports[`ClassicalElements should return true if the orbit is retrograde 2`] = `true`;
